### PR TITLE
Update tactics

### PR DESCRIPTION
**Pull request by PhiBabin**
- Remove old AiCommand everywhere and actions
- Fix almost all of the strategies 
- To actually remove unused actions
- Need *more testing* before merge

### CODE REVIEW

File: Engine/controller.py [hunk 1]
```diff
@@ -12,7 +12,11 @@
 from Util.engine_command import EngineCommand
 from Util.constant import PLAYER_PER_TEAM
 from Util.team_color_service import TeamColorService
-from Util.csv_plotter import Observer
+
+class Observer:
```
Comment (sim0504) [inline Engine/controller.py]: This make more sense, was my mistake. Thanks for fixing :P

File: Util/ai_command.py [hunk 2]
```diff
@@ -33,13 +36,14 @@ def __init__(self):
 
     def addMoveTo(self, target: Union[Pose, Position], cruise_speed: float=1, end_speed: float=0, ball_collision: bool=True):
         assert isinstance(target, (Pose, Position))
+        assert isinstance(target, Pose) and not isinstance(target.position, np.ndarray)
```
Comment (sim0504) [inline Util/ai_command.py]: this check is done twice

File: Util/ai_command.py [hunk 3]
```diff
@@ -67,4 +71,13 @@ def MoveTo(target: Union[Pose, Position], cruise_speed: float=1, end_speed: floa
     return CmdBuilder().addMoveTo(target, cruise_speed, end_speed, ball_collision).build()
 
 
+def GoBetween(position1: Position, position2: Position, target: Position, minimum_distance: [int, float]=0):
```
Comment (sim0504) [inline Util/ai_command.py]: using `float` in the typing make int valid by default.

File: ai/GameDomainObjects/field.py [hunk 4]
```diff
@@ -19,6 +21,97 @@ def ball(self):
     def constant(self):
         return self._constant
 
+    @constant.setter
+    def constant(self, field: Dict):
+        print(field)
```
Comment (sim0504) [inline ai/GameDomainObjects/field.py]: print alert!

File: ai/STA/Action/GoBehind.py [hunk 5]
```diff
@@ -1,21 +1,22 @@
...
 from ai.GameDomainObjects import Player
-from ai.STA.Action.Action import Action
-from ai.states.game_state import GameState
 
+RAYON_AVOID = 300  # (mm)
 
-# noinspection PyUnresolvedReferences
-class GoBehind(Action):
+def GoBehind(player: Player, position1: Position, position2: Position=None,
+             distance_behind: [int, float]=250,
```
Comment (sim0504) [inline ai/STA/Action/GoBehind.py]: only float is needed

File: ai/STA/Action/ProtectGoal.py [hunk 6]
```diff
@@ -1,72 +1,42 @@
...
-        exec(self): Retourne la pose où se rendre.
-    Attributs (en plus de ceux de Action):
-        player_id : L'identifiant du gardien.
-        is_right_goal : Un booléen indiquant si le but à protéger est celui de droite.
-        minimum_distance : La distance minimale qu'il doit y avoir entre le gardien et le centre du but.
-        maximum_distance : La distance maximale qu'il doit y avoir entre le gardien et le centre du but.
-    """
-    def __init__(self, game_state: GameState, player: Player, is_right_goal: bool=True,
+def ProtectGoal(game_state: GameState, player: Player, is_right_goal: bool=True,
                  minimum_distance: [int, float]=150 / 2, maximum_distance: [int, float, None]=None):
```
Comment (sim0504) [inline ai/STA/Action/ProtectGoal.py]: same here. also `maximum_distance: Optional[float]`

File: ai/STA/Strategy/prepare_kickoff_defense.py [hunk 7]
```diff
@@ -30,18 +30,18 @@ def __init__(self, p_game_state):
             else -GameState().const["CENTER_CENTER_RADIUS"]
 
         # Positions objectifs des joueurs
-        attack_top_position = Pose(GameState().const["FIELD_OUR_GOAL_X_EXTERNAL"] / 10,
+        attack_top_position = Pose.from_values(GameState().const["FIELD_OUR_GOAL_X_EXTERNAL"] / 10,
```
Comment (sim0504) [inline ai/STA/Strategy/prepare_kickoff_defense.py]: we should make a comment that this is hardcoded and bad. our robot will alway face at 0 degree

File: ai/STA/Tactic/bump.py [hunk 8]
```diff
@@ -72,12 +73,12 @@ def push_ball(self):
         player_to_target = target - player
         player_to_target = 0.5 * player_to_target / np.linalg.norm(player_to_target)
         speed_pose = Pose(Position.from_array(player_to_target))
-        return Move(self.game_state, self.player, speed_pose)
+        return MoveTo(speed_pose)
```
Comment (sim0504) [inline ai/STA/Tactic/bump.py]: this is so legacy, not sure if it's worth a refactor tho

File: ai/STA/Tactic/tactic.py [hunk 9]
```diff
@@ -62,7 +62,7 @@ def exec(self) -> AICommand:
         if isinstance(next_ai_command, Action):
             raise RuntimeError("Action are deprecaded use CmdBuilder")
         if not isinstance(next_ai_command, AICommand):
-            raise RuntimeError("A tactic MUST return an AICommand, not a {}".format(type(next_ai_command)))
+            raise RuntimeError("A tactic MUST return an AICommand, not a {}. It's {} the cuprid.".format(type(next_ai_command), self.current_state))
```
Comment (sim0504) [inline ai/STA/Tactic/tactic.py]: culprit?

File: Util/ai_command.py (hunk 2 above)
Comment (PhiBabin) [reply Util/ai_command.py]: Fix in fix/role_assignment

File: Util/geometry.py [hunk 10]
```diff
@@ -3,10 +3,16 @@
 import math as m
 import numpy as np
 
-from Util import Position
+from Util.pose import Pose
+from Util.position import Position
 from typing import cast, Sequence, List
 
 
+def get_angle_between_three_points(start: Position, mid: Position, end: Position):
```
Comment (sim0504) [inline Util/geometry.py]: -> float

File: ai/Algorithm/evaluation_module.py [hunk 11]
```diff
@@ -145,7 +170,7 @@ def trajectory_score(pointA, pointsB, obstacle):
 
 # noinspection PyPep8Naming,PyUnresolvedReferences
 def best_position_in_region(player, A, B):
-    # Retourne la position (dans un rectangle aux coins A et B) la mieux placée pour une passe
+    # Retourne la position (dans un rectangle aux monitor A et B) la mieux placée pour une passe
```
Comment (sim0504) [inline ai/Algorithm/evaluation_module.py]: monitor?

File: ai/STA/Action/GoBehind.py [hunk 12]
```diff
@@ -1,21 +1,22 @@
...
 from Util.geometry import wrap_to_pi
 from ai.GameDomainObjects import Player
-from ai.STA.Action.Action import Action
-from ai.states.game_state import GameState
 
+RAYON_AVOID = 300  # (mm)
 
-# noinspection PyUnresolvedReferences
-class GoBehind(Action):
+def GoBehind(player: Player, position1: Position, position2: Position=None,
```
Comment (sim0504) [inline ai/STA/Action/GoBehind.py]: None should be Optional[Position]

File: ai/STA/Action/GoBehind.py [hunk 13]
```diff
@@ -28,89 +29,55 @@ class GoBehind(Action):
...
+    else:
+        if math.sqrt((player_x-x)**2+(player_y-y)**2) < 50:
+            x -= math.cos(theta) * 2
+            y -= math.sin(theta) * 2
+        destination_position = Position(x, y)
+
+    # Calcul de l'orientation de la pose de destination
+    destination_orientation = 0
+    if orientation == 'front':
+        destination_orientation = (position1 - destination_position).angle
```
Comment (sim0504) [inline ai/STA/Action/GoBehind.py]: this should wrap to pi

File: ai/STA/Tactic/align_to_defense_wall.py [hunk 14]
```diff
@@ -173,18 +174,17 @@ def exec(self):
         else:
             destination_orientation = (self.ball_position -
                                        self.positions_in_formations[self.player_number_in_formation]).angle
-            return GoToPositionPathfinder(self.game_state, self.player,
-                                          Pose(self.positions_in_formations[self.player_number_in_formation],
-                                               destination_orientation)).exec()
+            return CmdBuilder().addMoveTo(Pose(self.positions_in_formations[self.player_number_in_formation],
```
Comment (sim0504) [inline ai/STA/Tactic/align_to_defense_wall.py]: could be a simple MoveTo(...)

File: ai/STA/Tactic/align_to_defense_wall.py [hunk 15]
```diff
@@ -173,18 +174,17 @@ def exec(self):
...
+                                               destination_orientation)).build()
 
     def halt(self):
         self.status_flag = Flags.SUCCESS
-        return Idle(self.game_state, self.player)
+        return Idle
```
Comment (sim0504) [inline ai/STA/Tactic/align_to_defense_wall.py]: should be Idle() I think

File: ai/STA/Tactic/demo_follow_robot.py [hunk 16]
```diff
@@ -23,20 +22,20 @@ def __init__(self, game_state: GameState, player: Player, p_target: Pose=Pose(),
 
     def move_to_ball(self):
         self.status_flag = Flags.WIP
-        self.target = self.game_state.get_player_pose(self.robot_to_follow_id)
+        self.target = self.game_state.get_player(self.robot_to_follow_id).pose
```
Comment (sim0504) [inline ai/STA/Tactic/demo_follow_robot.py]: you can access .position directly to avoid indirection.

File: ai/STA/Tactic/demo_follow_robot.py [hunk 17]
```diff
@@ -23,20 +22,20 @@ def __init__(self, game_state: GameState, player: Player, p_target: Pose=Pose(),
...
 
         if (self.player.pose.position - self.target.position).norm < POSITION_DEADZONE + ROBOT_RADIUS:
             self.next_state = self.halt
         else:
             self.next_state = self.move_to_ball
 
-        return MoveToPosition(self.game_state, self.player, self.target)
+        return CmdBuilder().addMoveTo(self.target).build()
```
Comment (sim0504) [inline ai/STA/Tactic/demo_follow_robot.py]: could be MoveTo(...)

File: ai/STA/Tactic/demo_follow_robot.py [hunk 18]
```diff
@@ -23,20 +22,20 @@ def __init__(self, game_state: GameState, player: Player, p_target: Pose=Pose(),
...
     def halt(self):
         self.status_flag = Flags.SUCCESS
 
-        if (self.player.pose.position - self.game_state.ball_position) < POSITION_DEADZONE + ROBOT_RADIUS:
+        if (self.player.pose.position - self.game_state.ball_position).norm < POSITION_DEADZONE + ROBOT_RADIUS:
             self.next_state = self.halt
         else:
             self.next_state = self.move_to_ball
-        return Idle(self.game_state, self.player)
+        return Idle
```
Comment (sim0504) [inline ai/STA/Tactic/demo_follow_robot.py]: Idle()

File: ai/STA/Tactic/do_kick.py [hunk 19]
```diff
@@ -21,4 +21,4 @@ def __init__(self, game_state: GameState, player: Player, target: Pose=Pose(), a
 
     def kick(self):
         self.next_state = self.halt
-        return Kick(self.game_state, self.player, self.kick_force)
+        return CmdBuilder().addKick(kick_force=self.kick_force).build()
```
Comment (sim0504) [inline ai/STA/Tactic/do_kick.py]: is there a shorthand for kick? like `Kick(kick_force-...)` ?

File: ai/STA/Tactic/go_to_random_pose_in_zone.py [hunk 20]
```diff
@@ -47,10 +49,10 @@ def exec(self):
             self.next_pose = Pose(self.current_position_to_go, self.current_angle_to_go)
         self.next_state = self.exec
 
-        return GoToPositionPathfinder(self.game_state, self.player, self.next_pose).exec()
+        return CmdBuilder().addMoveTo(self.next_pose).build()
```
Comment (sim0504) [inline ai/STA/Tactic/go_to_random_pose_in_zone.py]: MOveTo()

File: ai/STA/Tactic/goalkeeper.py [hunk 21]
```diff
@@ -90,17 +85,17 @@ def protect_goal(self):
                 width = self.game_state.const["FIELD_GOAL_WIDTH"]
                 y_position_on_line = clamp(y_position_on_line, -width, width)
 
-                destination = Pose(our_goal.x, y_position_on_line, goalkeeper_orientation)
+                destination = Pose(Position(our_goal.x, y_position_on_line), goalkeeper_orientation)
```
Comment (sim0504) [inline ai/STA/Tactic/goalkeeper.py]: Pose.from_values(our_goal.x, y_position_on_line, goalkeeper_orientation)

File: ai/STA/Tactic/goalkeeper.py [hunk 22]
```diff
@@ -90,17 +85,17 @@ def protect_goal(self):
...
 
             else:
                 destination = Pose(our_goal)
-            return MoveToPosition(self.game_state, self.player, destination, pathfinder_on=True, cruise_speed=2)
+            return CmdBuilder().addMoveTo(destination, cruise_speed=2).build()
```
Comment (sim0504) [inline ai/STA/Tactic/goalkeeper.py]: MoveTo(...)

File: ai/STA/Tactic/goalkeeper.py [hunk 23]
```diff
@@ -110,9 +105,9 @@ def go_behind_ball(self):
         else:
             self.next_state = self.go_behind_ball
             self._find_best_passing_option()
-        collision_ball = self.tries_flag == 0
-        return GoToPositionPathfinder(self.game_state, self.player, Pose(distance_behind, orientation),
-                                      collision_ball=collision_ball, cruise_speed=2, end_speed=0.2)
+        ball_collision = self.tries_flag == 0
+        return CmdBuilder().addMoveTo(Pose(distance_behind, orientation),
```
Comment (sim0504) [inline ai/STA/Tactic/goalkeeper.py]: MoveTo(...)

File: ai/STA/Tactic/intercept.py [hunk 24]
```diff
@@ -105,4 +104,4 @@ def _is_player_towards_ball(self, fact=-0.99):
 
     def halt(self):
         self.status_flag = Flags.SUCCESS
-        return Idle(self.game_state, self.player)
+        return Idle
```
Comment (sim0504) [inline ai/STA/Tactic/intercept.py]: Idle()

File: ai/STA/Tactic/pass_to_player.py [hunk 25]
```diff
@@ -73,30 +68,30 @@ def kick(self):
             self.next_state = self.kick
         else:
             self.next_state = self.kick_charge
-        return Kick(self.game_state, self.player, 5, self.target)
+        return CmdBuilder().addKick(KickForce.HIGH).addMoveTo(self.target).build()
```
Comment (sim0504) [inline ai/STA/Tactic/pass_to_player.py]: Kick(...) like the other one

File: ai/STA/Tactic/pass_to_player.py [hunk 26]
```diff
@@ -73,30 +68,30 @@ def kick(self):
...
 
     def halt(self):
         if self.status_flag == Flags.INIT:
             self.next_state = self.kick_charge
         else:
             self.status_flag = Flags.SUCCESS
-        return Idle(self.game_state, self.player)
+        return Idle
```
Comment (sim0504) [inline ai/STA/Tactic/pass_to_player.py]: i think this is right finally, sorry for the confusion :P

File: ai/STA/Tactic/position_for_pass.py [hunk 27]
```diff
@@ -24,7 +25,7 @@
 class PositionForPass(Tactic):
 
     def __init__(self, game_state: GameState, player: Player, target: Pose=Pose(), args: List[str]=None,
-                 auto_position=False):
+                 auto_position=False, robots_in_formation: List[Player] = None):
```
Comment (sim0504) [inline ai/STA/Tactic/position_for_pass.py]: should be: `... args: Optional[List[str]]=None, auto_position: bool=False, robots_in_formation: Optional[List[Player]]=None`

File: ai/STA/Tactic/position_for_pass.py [hunk 28]
```diff
@@ -54,14 +55,15 @@ def is_player_offense(self, player):
     def move_to_pass_position(self):
 
         self.next_state = self.move_to_pass_position
-        return GoToPositionPathfinder(self.game_state, self.player, self._get_destination_pose())
+        return CmdBuilder().addMoveTo(self._get_destination_pose()).build()
```
Comment (sim0504) [inline ai/STA/Tactic/position_for_pass.py]: MoveTo(...)

File: ai/STA/Tactic/stay_away_from_ball.py [hunk 29]
```diff
@@ -1,29 +1,26 @@
...
 from ai.GameDomainObjects import Player
-from ai.STA.Tactic.go_to_position_pathfinder import GoToPositionPathfinder
 from ai.STA.Tactic.tactic import Tactic
 from ai.states.game_state import GameState
 
 
 class StayAwayFromBall(Tactic):
     def __init__(self, game_state: GameState, player: Player, target: Pose = Pose(),
-                 keepout_radius: int = 500, args: List[str]=None):
+                 args: List[str]=None, keepout_radius: int = 500):
```
Comment (sim0504) [inline ai/STA/Tactic/stay_away_from_ball.py]: Optional[List[str]]

Comment (sim0504) [reply ai/STA/Tactic/stay_away_from_ball.py]: int should be float

File: config/config.py [hunk 30]
```diff
@@ -117,4 +115,11 @@ def __getitem__(self, item):
     def __setitem__(self, key, value):
         if self._config_was_set:
             raise RuntimeError('You can\'t change the configuration after it has been loaded from a file.')
-        self._config[key] = value
\ No newline at end of file
+        self._config[key] = value
+
+    def load_parameters(self, cli_args):
```
Comment (sim0504) [inline config/config.py]: we could use the update method from the dict for each section. This avoid the hardcoding of the key.

self._config['ENGINE'].update(engine_args)
self._config['GAME].update(game_args)

File: ai/STA/Tactic/demo_follow_robot.py (hunk 16 above)
Comment (PhiBabin) [reply ai/STA/Tactic/demo_follow_robot.py]: But we don't need the position, we need the pose

**Pull request merged**